### PR TITLE
25076 Add algorithm for clipping peaks

### LINKED ISSUE #25076: Bring SNAP_Background_Peak_Clipping into mantid
Create peak striping algorithm from `/SNS/SNAP/shared/Mantid_algorithms/SNAP_Background_Peak_Clipping.py` and refactor `SNAPReduce` to use the new algorithm. The new algorithm should have a more generic name.

### PR DESCRIPTION
**Pull request by ckendrick**
**Description of work.**

Implements the SNAPBackgroundPeakClipping algorithm provided in the referenced issue as a new algorithm 'ClipPeaks' which can be used to remove peaks from data. The majority of the algorithm already existed in 'SNAPReduce,' so the functions in the original file were updated to the newer ones found in SNAPReduce. The SNAPReduce algorithm will be refactored to call this new algorithm in a separate issue/pull request.

Documentation for the algorithm was added, including the equations the algorithm is based on. Some basic unit tests were created from adding peaks to a simple baseline function and verifying the ClipPeaks algorithm removed the peaks and was able to recreate a close version of the baseline function.

**To test:**

Download the attached script and open in Mantid: [clippeak_test.py.txt](https://github.com/mantidproject/mantid/files/5220415/clippeak_test.py.txt)

Once the script is executed, several workspaces will be created:
The "BaselineData" workspace can be plotted to show what the original function is.
"NoiseData" is the baseline function with randomly generated peaks added.

The tester should plot the spectrum of "PeakClippedData" and verify that the peaks randomly introduced in "NoiseData" are gone, and that it resembles the original "BaselineData" plot. 

The tester can also re-run the script several times, verifying the above result remains true for different random seeds.

Resolves #25076.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ckendrick** [comment]
Ubuntu + Doc Tests failed at "MantidQtScientificInterfacesTestQt4_MuonAnalysisResultTableCreatorTest" which was not modified by this PR.

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        weighted 4, i=/- 1 weighted 3, i+/-2 weighted 2 and i+/-3
+        weighted 1 this input is only the y values
+        """
+        sm = np.zeros(len(data))
+        factor = int(order / 2) + 1
+
+        for i in range(len(data)):
+            temp = 0
+            ave = 0
+            for r in range(max(0, i - int(order / 2)), min(i + int(order / 2), len(data) - 1) + 1):
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:85]: Using numpy array operation to replace for loop: (the following might not be correct as I haven't tested)

```
min_index = max(0, i - int(order / 2))
max_index = min(i + int(order / 2), len(data) - 1)
distance = np.arange(max_index - min_index) - i
temp_sum = np.sum(factor - np.abs(distance)) * data[min_index: max_index])
temp_avg = np.sum(factor - np.abs(distance))
```

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        Performs peak clipping on the data iteratively over the size of win. The input data can be optionally
+        smoothed if smooth_window is greater than 0. The LLS function can be performed on the data after smoothing
+        to help exaggerate smaller peaks present in the data before clipping. The peak clipping window used for
+        iteration is done with a decreasing window unless decrease is False.
+
+        :param data: Input (Y) data to clip.
+        :param win: Window size of the clipping (controls the iteration count of the clipping function).
+        :param decrease: Flag to determine if an increasing/decreasing window is applied to the data.
+        :param lls: Flag to enable/disable applying the LLS function to data before clipping.
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:119]: rename to `use_lls`?

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    def peak_clip(self, data, win=30, decrease=True, lls=True, smooth_window=0):
+        """
+        Performs peak clipping on the data iteratively over the size of win. The input data can be optionally
+        smoothed if smooth_window is greater than 0. The LLS function can be performed on the data after smoothing
+        to help exaggerate smaller peaks present in the data before clipping. The peak clipping window used for
+        iteration is done with a decreasing window unless decrease is False.
+
+        :param data: Input (Y) data to clip.
+        :param win: Window size of the clipping (controls the iteration count of the clipping function).
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:117]: rename to `clip_window_size`?

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        :param decrease: Flag to determine if an increasing/decreasing window is applied to the data.
+        :param lls: Flag to enable/disable applying the LLS function to data before clipping.
+        :param smooth_window: Smoothing factor for the smoothing function. 0 to apply no smoothing.
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:120]: rename to `smooth_factor`?

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        window = win
+        self.log().information('Smoothing window: {0}'.format(str(smooth_window)))
+
+        if smooth_window > 0:
+            data = self.smooth(data, smooth_window)
+
+        if lls:
+            data = self.log_log_sqrt_transformation(data)
+
+        temp = data.copy()
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:134]: Is it necessary to make a copy of **data**, if **data** is not used in this method anymore?

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+        if decrease:
+            scan = range(window + 1, 0, -1)
+        else:
+            scan = range(1, window + 1)
+
+        for w in scan:
+            for i in range(len(temp)):
+                if i < w or i > (len(temp) - w - 1):
+                    continue
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:144]: Please add comment on why 'continue' here.

File: Framework/PythonInterface/plugins/algorithms/ClipPeaks.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+        n_histo = input_ws.getNumberHistograms()
+
+        peak_clip_ws = CloneWorkspace(input_ws, OutputWorkspace=self.getPropertyValue("OutputWorkspace"))
+
+        x = input_ws.extractX()
+        y = input_ws.extractY()
+        e = input_ws.extractE()
+
+        for h in range(n_histo):
```
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/ClipPeaks.py:191]: It is better to use `histogram` than `h` for readability.

**Review by wdzhou — APPROVED** [review]
It looks good to me!  :shipit:

**Pull request merged**